File: Linked_List/138. Copy List with Random Pointer.py
class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if not head:
            return None
        m = dict()
        m[head] = Node(head.val)
        cur = head        
        while cur and cur.next:
            m[cur.next] = Node(cur.next.val)
            cur = cur.next
        cur = head
        while cur:
            if cur.random:
                m[cur].random = m[cur.random]
            if cur.next:
                m[cur].next = m[cur.next]
            cur = cur.next
        return m[head]

File: Linked_List/138. Copy List with Random Pointer.cpp
// 跟copy graph一样

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if(head == nullptr)
            return nullptr;
        unordered_map<Node*, Node*> m;
        m[head]  = new Node(head->val);
        Node* cur = head;
        while(cur && cur->next)
        {
            m[cur->next] = new Node(cur->next->val);
            m[cur]->next = m[cur->next];
            cur = cur->next;
        }
        cur = head;
        while(cur)
        {
            if(cur->random != nullptr)
                m[cur]->random = m[cur->random];
            cur = cur->next;
        }
        return m[head];
    }
};
